cmd/installer/cli/upgrade.go

```diff
@@ -0,0 +1,778 @@
+			return fmt.Errorf("unable to read license file: %w", err)
+		}
+		flags.licenseBytes = b
+
+		// validate the license is indeed a license file
+		l, err := helpers.ParseLicense(flags.licenseFile)
+		if err != nil {
+			if err == helpers.ErrNotALicenseFile {
+				return fmt.Errorf("license file is not a valid license file")
+			}
+			return fmt.Errorf("unable to parse license file: %w", err)
+		}
+		flags.license = l
+	}
+
+	if flags.configValues != "" {
+		err := configutils.ValidateKotsConfigValues(flags.configValues)
+		if err != nil {
+			return fmt.Errorf("config values file is not valid: %w", err)
+		}
+	}
+
+	flags.isAirgap = flags.airgapBundle != ""
+	if flags.airgapBundle != "" {
+		metadata, err := airgap.AirgapMetadataFromPath(flags.airgapBundle)
+		if err != nil {
+			return fmt.Errorf("failed to get airgap info: %w", err)
+		}
+		flags.airgapMetadata = metadata
+	}
+
+	var err error
+	flags.embeddedAssetsSize, err = goods.SizeOfEmbeddedAssets()
+	if err != nil {
+		return fmt.Errorf("failed to get size of embedded files: %w", err)
+	}
+
+	// Read existing TLS certificates from cluster secrets
+	// This is required for upgrades - we must use certificates from the existing installation
+	if err := readKotsadmTLSSecret(flags); err != nil {
+		return fmt.Errorf("upgrade requires TLS certificates from existing installation: %w", err)
+	}
+
+	// Read existing admin console configuration from cluster secrets
+	if err := readKotsadmConfigMap(flags); err != nil {
+		return fmt.Errorf("failed to read admin console configuration from existing installation: %w", err)
+	}
+
+	if flags.managerPort != 0 && flags.adminConsolePort != 0 {
+		if flags.managerPort == flags.adminConsolePort {
+			return fmt.Errorf("manager port cannot be the same as admin console port")
+		}
+	}
+
+	proxy, err := proxyConfigFromCmd(cmd, flags.assumeYes)
+	if err != nil {
+		return err
+	}
+
+	rc.SetAdminConsolePort(flags.adminConsolePort)
+	ki.SetAdminConsolePort(flags.adminConsolePort)
+
+	rc.SetManagerPort(flags.managerPort)
+	ki.SetManagerPort(flags.managerPort)
+
+	rc.SetProxySpec(proxy)
+	ki.SetProxySpec(proxy)
+
+	// Process TLS certificate configuration if provided
+	if err := processUpgradeTLSConfig(flags); err != nil {
+		return fmt.Errorf("process TLS configuration: %w", err)
+	}
+
+	return nil
+}
+
+func preRunUpgradeLinux(cmd *cobra.Command, flags *UpgradeCmdFlags, rc runtimeconfig.RuntimeConfig) error {
+	if !cmd.Flags().Changed("ignore-host-preflights") && (os.Getenv("SKIP_HOST_PREFLIGHTS") == "1" || os.Getenv("SKIP_HOST_PREFLIGHTS") == "true") {
+		flags.ignoreHostPreflights = true
+	}
+
+	if os.Getuid() != 0 {
+		return fmt.Errorf("upgrade command must be run as root")
+	}
+
+	// set the umask to 022 so that we can create files/directories with 755 permissions
+	// this does not return an error - it returns the previous umask
+	_ = syscall.Umask(0o022)
+
+	// If data directory wasn't provided, try to detect it from existing installation
```

Comment:
user shouldn't be able to provide a data dir in the upgrade command. in fact, we should disable some fields that can't be updated in the setup page. not in scope for this pr though